w3ctrace/state.go

```diff
@@ -17,109 +18,116 @@ const maxKVPairs = 32
+	filteredStates := listMembers[:0]
+	i := 0
+	for ; itemsToFilter > 0 && i < len(listMembers); i++ {
+		if len(listMembers[i]) > thresholdLen {
+			itemsToFilter--
+			continue
+		}
+		filteredStates = append(filteredStates, listMembers[i])
 	}
+	filteredStates = append(filteredStates, listMembers[i:]...)
 
-	return strings.TrimPrefix(st[i], vendor+"="), true
-}
-
-// Index returns the index of vendor-specific data for given vendor in the state.
-// It returns -1 if the state does not contain data for this vendor.
-func (st State) Index(vendor string) int {
-	prefix := vendor + "="
-
-	for i, vd := range st {
-		if strings.HasPrefix(vd, prefix) {
-			return i
-		}
+	if len(filteredStates) > maxListMembers {
+		return State{listMembers: filteredStates[:maxListMembers], instanaTraceStateValue: instanaTraceStateValue}, nil
 	}
 
-	return -1
+	return State{listMembers: filteredStates, instanaTraceStateValue: instanaTraceStateValue}, nil
 }
 
-// Remove returns a new state without data for specified vendor. It returns the same state if vendor is empty
-func (st State) Remove(vendor string) State {
-	if vendor == "" {
-		return st
+// SetInstanaValue returns a new state prepended with the provided Instana value. If the original state had an Instana
+// list member pair, it is discarded/overwritten.
+func (st State) SetInstanaTraceStateValue(instanaTraceStateValue string) State {
+	var listMembers []string
+	if st.instanaTraceStateValue == "" && instanaTraceStateValue != "" && len(st.listMembers) == maxKVPairs {
+		// The incoming tracestate had the maximum number of list members but no Instana list member, now we are adding an
+		// Instana list member, so we would exceed the maximum by one. Hence, we need to discard one of the other list
+		// members to stay within the limits mandated by the specification.
+		listMembers = st.listMembers[:maxKVPairs-1]
+	} else {
+		listMembers = st.listMembers
 	}
 
-	prefix := vendor + "="
-
-	var newSt State
-	for _, vd := range st {
-		if !strings.HasPrefix(vd, prefix) {
-			newSt = append(newSt, vd)
-		}
-	}
+	return State{listMembers: listMembers, instanaTraceStateValue: instanaTraceStateValue}
+}
 
-	return newSt
+// FetchInstanaTraceStateValue retrieves the value of the Instana tracestate list member, if any.
+func (st State) FetchInstanaTraceStateValue() (string, bool) {
+	return st.instanaTraceStateValue, st.instanaTraceStateValue != ""
 }
 
 // String returns string representation of a trace state. The returned value is compatible with the
-// `tracestate` header format
+// `tracestate` header format. If the state has an Instana-specific list member, that one is always rendered first. This
+// is optimized for the use case of injecting the string represenation of the tracestate header into downstream
+// requests.
 func (st State) String() string {
-	if len(st) == 0 {
+	if len(st.listMembers) == 0 && st.instanaTraceStateValue == "" {
 		return ""
+	} else if len(st.listMembers) == 0 {
```

Comment:
A common practice of the Go Team is to use guard clauses. 
So instead of using `else if`, could you please start a brand new if condition?

Reply:
Done.